Allows the Parser to handle event files as well.

schnika opened the pull request:
!WIP! 

* I don't like the `calendars_or_events` variable.
* I don't like the fact that the very same file may contain multiple calendars and events.

I like the fact, that the interface stays the same. E.g. `Icalendar.parse` works for both events and calendars. However it should be clear, that you can parse either calendar or event files. 

Regarding the `Icalendar::Parser`:

What do you think about a guard in the parser which makes sure that events and calendars can't be mixed?

What do you think about extending the interface to sth. like:

`Icalendar::Parser.new(:calendar, source)`

or 

`Icalendar::Parser.new(:event, source)`

STI would be an option as well `Icalendar::EventParser.new(source)` or `Icalendar::CalendarParser.new(source)` where both parsers inherit from `Icalendar::Parser.new` and override/extend the parse method.

This way it would be a lot clearer what we expect from the source and what it will return... We would get rid of the `calendars_or_events` variable and it would be nicely decoupled?

Please tell me what you think :)

rahearn commented:
Solid first effort. I think I share all of your same concerns about this solution. I also think that no matter which way we go, we should maintain backwards compatibility, so the method signatures and return values of `Icalendar.parse` and `Icalendar::Parser.new` can't change, though we can add new optional parameters to them.

What if we modified the parser to handle bare events by wrapping them in an otherwise nonexistent calendar? That way, the parser always returns one or more `Icalendar::Calendar` models. It would allow for a source file with mixes of calendars and events, while the consuming code could treat it just like a file with multiple calendars.

schnika commented:
I totally agree on maintaining the backwards compatibility. 

However, without having a strong opinion against it, I don't like the idea of supporting mixed events and calendars in one file. This could lead to a whole lot of confusion/errors in the future, e.g. which events in which calendar (merge them or don't, what is separating them, ...)

I will push a new draft in a couple of minutes

schnika commented:
With 4e9de93 the `Parser.parse` method is a bit more generic. You can tell the parser which component he is expecting (calendar by default in order to maintain backwards compatibility) This makes it easy to enhance the parser to handle not only events but all other components as well.

Adds the `Component.parse` method. 
This way `Icalendar::Event.parse` or `Icalendar::Calendar.parse` is working aside from `Icalendar.parse`. I think `Icalendar.parse` could get a deprecation warning as well because the usage of Icalendar::{Component}.parse is more intuitive than calling `parse` on the 'base class'.  With the provided implementation the user gets no surprises :) 

What do you think?

This is still a draft. I am thinking about validating if the component is supported for parsing already on the component level e.g. like this:

`````
class Alarm < Component
  def self.parse
    raise ComponentNotParseableError.new, "..."
  end
end
`````

Or simply not defining `Component.parse` but explicitly have it ONLY on `Event` and `Calendar`. However I think it would be nice to have this feature for all Components in the future.

schnika commented on README.md:
```diff
@@ -273,6 +273,20 @@ strict parsing:
     strict_parser = Icalendar::Parser.new(cal_file, true)
     cal = strict_parser.parse
 
+Parsing Events
```
This is not right anymore

schnika commented:
@rahearn do you have any update for me? I would love to finish this feature :)

rahearn commented:
@schnika sorry it's taken me so long to get back to this.

I don't see any reason why any component wouldn't be parseable with the way you have it set up now. I think we can just removed the `parseable?` methods and allow everything that inherits from `Component` to parse, unless you know of a good reason to not allow that?

I'm fine with adding a deprecation warning to `Icalendar.parse`

schnika commented:
@rahearn thanks for your reply. I was a bit too passive with my approach. I added the deprecation warning, and updated the docs accordingly. If you have anything else for me, tell me :)

schnika commented on spec/event_spec.rb:
```diff
@@ -96,6 +96,18 @@
     end
   end
 
+  describe '.parse' do
```
this is already covered through the parser_spec. could remove it for DRYness reasons...

rahearn replied:
I don't think it harms anything to leave it in. Should be fast enough and verifies that we set up the `Parser#component` properly.

rahearn commented on README.md:
```diff
@@ -273,6 +273,20 @@ strict parsing:
+    event_file = File.open("event.ics")
+
+    # Parser returns an array of events because a single file
+    # can have multiple calendars.
```
can have multiple ~~calendars~~ events

rahearn commented on lib/icalendar.rb:
```diff
@@ -13,6 +13,7 @@ def self.logger=(logger)
   end
 
   def self.parse(source, single = false)
+    puts "**** DEPRECATION WARNING ****\nIcalender.parse will be removed. Please switch to Icalendar::Calendar.parse."
```
Please use `warn` instead of `puts` for the deprecation warning.

rahearn commented:
@schnika Please address the 2 comments and see what you can do about TravisCI failing. I don't want to merge this as long as that is failing. Otherwise this looks good to me.

schnika commented:
@rahearn back to you. Tests are green now + I updated the `roundtrip_spec.rb` to use the promoted syntax for importing components. 
I added ruby 2.2 and ruby 2.3.0 to the travis ci build list as well. So we are up to date there as well. I tried to fix the ruby-head build but there still seems to be an issue:
https://github.com/bundler/bundler/pull/3559
